### ISSUE TITLE
Communication between Kubernetes components

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#6363]
Conversation started by @davidopp due to complexities arising from node-to-master communication (e.g., #6193, #6285, #6077, #6063, #6052, #5953).

Overall communication issues:
- performance, latency, scalability
- DoS
- RAW and WAW consistency
- clock skew / time synchronization
- true and false conflicts
- auth
- component discovery

I'll wait for @davidopp to fill in his proposal.

cc @erictune @lavalamp @thockin @dchen1107 @smarterclayton @gmarek

**Comment by gmarek** [comment]
+1 to anything that will make the reasoning simple. Current state when multiple component can write exactly the same thing make reasoning about system semantics very hard (to put it lightly).

**Comment by smarterclayton** [comment]
cc @liggitt

**Comment by bgrant0607** [comment]
And types of communication:

To node:
* Node/Kubelet configuration
* Pods to run and their desired states
* Services and endpoints
* Secrets
* Persistent volumes

From node:
* System info (software versions, etc.)
* Pod, container, and volume statuses
* Node heart beats (Kubelet alive)
* Node health status (docker working, disks ok, etc.)
* Resource usage (#4057)

**Comment by davidopp** [comment]
Thanks for filing the issue. Give me two seconds before you start discussing it, so that I can actually write it up. :)

**Comment by zmerlynn** [comment]
cc @zmerlynn, and I suspect @roberthbailey for SSLness

**Comment by smarterclayton** [comment]
* Node bootstrap (distributing client certs to the nodes)

**Comment by bgrant0607** [comment]
* Cluster (master and node) upgrade

**Comment by bgrant0607** [comment]
* "Free range" nodes (no master), such as ContainerVM

**Comment by davidopp** [comment]
[Note: I am likely to update this text as I think of more things, so please read this issue through the web, not email.]

I think we need to talk about how we structure communication between components in Kubernetes. I'm not necessarily proposing a change pre-1.0; this is more of a longer-term idea.

Currently, all communication between components is through objects that are (1) persistent, and (2) considered to be part of the external API, which is stable/slow-evolving, versioned with long deprecation periods, and subject to multiple levels of review, debate, and approval for changes. I believe this one-size-fits-all model is bad for performance/scalability and bad for the project's velocity.

I'm proposing an additional, lighter-weight model inter-component communication mechanism, which has neither property (1) nor (2), to be used in situations where the existing model is deemed inappropriate. (Exactly what those circumstances are is TBD, but it would *only* be used for some subset of the internal communication between Kubernetes components.) This model is a very small tweak to the existing model. The only difference is that the objects it uses would not be persisted. We would exploit this fact to allow APIs that use these objects to evolve more quickly. But everything would still go through PUT/POST/GET to the API server.

Let me give you a concrete example of how it would work. Let's say we move NodeStatus to this model. Let's also assume the API server is sharded, to alleviate concerns that this approach doesn't work with sharding. A node would POST/PUT a NodeStatus object to its API server shard. The API server would do whatever business logic processing it wants, and if there is some state that all of the API server replicas need to know, and/or that is needed for crash recovery, it would write a new object with just that information to etcd. For example, we might have a UserVisibleNodeStatus object that is the one external clients read; it contains a processed subset of the information from NodeStatus (and perhaps from other objects as well) and can be updated less frequently.

(Please don't dwell too much on my use of NodeStatus as the example. I think this approach applies to inter-component communication in Kubernetes more generally.)

Here are the advantages I see
* performance/scalability: The current model, where components can only communicate with one another through disk, is an obvious performance bottleneck. Offloading some of this communication to network+memory-only paths is an obvious performance/scalability win.
* state partitioning: The current model conflates objects that are needed for inter-component communication with objects that are needed for crash recovery. Only the second type need to be persisted, but we're forced to persist everything.
* API layering: The current model doesn't distinguish internal APIs from external APIs. I know some people think there shouldn't actually be a distinction, but I disagree. If we can declare some APIs/API objects to be internal-only and never persisted, then we do not have to be as rigorous about versioning, long deprecation periods, evolving slowly/carefully, etc. for these APIs (of course there are caveats here about atomicity of component upgrade for loosening the versioning requirements). The messages used to communicate between components could be completely invisible to the user, with the API server responsible for generating the objects that are part of the user-visible API. Components could have input and output objects reminiscent of RPCs without having to worry about how they affect the user-facing API

The only gotcha I can see is that we'd need to implement some kind of replacement for etcd watch, to allow components to watch these ephemeral objects. I think we wouldn't need something like ResourceVersion if we assume each of these ephemeral objects only have one writer.

**Comment by smarterclayton** [comment]
> On Apr 2, 2015, at 4:51 PM, David Oppenheimer <notifications@github.com> wrote:
> 
> [Note: I am likely to update this text as I think of more things, so please read this issue through the web, not email.]
> 
> I think we need to talk about how we structure communication between components in Kubernetes. I'm not necessarily proposing a change pre-1.0; this is more of a longer-term idea.
> 
> Currently, all communication between components is through objects that are (1) persistent, and (2) considered to be part of the external API, which is stable/slow-evolving, versioned with long deprecation periods, and subject to multiple levels of review, debate, and approval for changes. I believe this one-size-fits-all model is bad for performance/scalability and bad for the project's velocity.
> 
> I'm proposing an additional, lighter-weight model inter-component communication mechanism, which has neither property (1) nor (2), to be used in situations where the existing model is deemed inappropriate. (Exactly what those circumstances are is TBD, but it would only be used for some subset of the internal communication between Kubernetes components.) This model is a very small tweak to the existing model. The only difference is that the objects it uses would not be persisted. We would exploit this fact to allow APIs that use these objects to evolve more quickly. But everything would still go through PUT/POST/GET to the API server.
> 
> Let me give you a concrete example of how it would work. Let's say we move NodeStatus to this model. Let's also assume the API server is sharded, to alleviate concerns that this approach doesn't work with sharding. A node would POST/PUT a NodeStatus object to its API server shard. The API server would do whatever business logic processing it wants, and if there is some state that all of the API server replicas need to know, and/or that is needed for crash recovery, it would write a new object with just that information to etcd. For example, we might have a UserVisibleNodeStatus object that is the one external clients read; it contains a processed subset of the information from NodeStatus (and perhaps from other objects as well) and can be updated less frequently.
> 
> (Please don't dwell too much on my use of NodeStatus as the example. I think this approach applies to inter-component communication in Kubernetes more generally.)
> 
> Here are the advantages I see
> 
> performance/scalability: The current model, where components can only communicate with one another through disk, is an obvious performance bottleneck. Offloading some of this communication to network+memory-only paths is an obvious performance/scalability win.
> state partitioning: The current model conflates objects that are needed for inter-component communication with objects that are needed for crash recovery. Only the second type need to be persisted, but we're forced to persist everything.
> API layering: The current model doesn't distinguish internal APIs from external APIs. I know some people think there shouldn't actually be a distinction, but I disagree. If we can declare some APIs/API objects to be internal-only and never persisted, then we do not have to be as rigorous about versioning, long deprecation periods, evolving slowly/carefully, etc. for these APIs (of course there are caveats here about atomicity of component upgrade for loosening the versioning requirements). The messages used to communicate between components could be completely invisible to the user, with the API server responsible for generating the objects that are part of the user-visible API. Components could have input and output objects reminiscent of RPCs without having to worry about how they affect the user-facing API
To be honest, no matter how you specify the APIs, people will use them if the need them to solve the same use case, and then be upset when they break.  So we should be careful to identify when that happens and address the need.

I don't see an issue with evolving internal APIs at a different pace and with different guarantees.  Virtual resources makes a lot of sense when we want to specialize the use case.

**Comment by lavalamp** [comment]
> The only difference is that the objects it uses would not be persisted. 

I am not sure I see the point if that is the only difference. All the heavyweight process is also for maintaining API compatibility, which you still need-- clusters have to upgrade.

**Comment by smarterclayton** [comment]
When someone says "heavyweight process" w.r.t. API design, I mentally substitute "discipline" :)

**Comment by lavalamp** [comment]
> API server is sharded

It might be replicated, but that's not sharding. I would propose that the storage layer (etcd today) is the place where we do sharding, should that be necessary. Otherwise every client needs to be aware of the sharding details, or we have to put in another layer.

**Comment by lavalamp** [comment]
> performance/scalability: The current model, where components can only communicate with one another through disk, is an obvious performance bottleneck. ...

This is an argument for changing the characteristics of our storage backend, IMO.

> state partitioning: The current model conflates objects that are needed for inter-component communication with objects that are needed for crash recovery. Only the second type need to be persisted, but we're forced to persist everything.

Not true. See our binding object, which is not persisted. (directly)

> API layering: The current model doesn't distinguish internal APIs from external APIs. ... If we can declare some APIs/API objects to be internal-only and never persisted, then we do not have to be as rigorous about versioning, long deprecation periods, evolving slowly/carefully, etc. for these APIs ...

I guess I just don't see how that would actually cash out in a simplified compatibility matrix.

**Comment by dchen1107** [comment]
[Note: copied from internal email with some modification]

@davidopp I agreed with you on requiring every message to be persisted is an issue for our performance and scalability (#5953). On kubelet side, to cope with the rule and avoid too much writes, kubelet only post PodStatus on change (which is good anyway). But we use NodeStatus as ping message today to indicate if a node alive, which is quite expensive since it requires posting in a high
frequency. It was hard to me to accept persisting a ping message to store.

Related to today's issue caused by NodeStatus and NodeController, there are several ways potentially to solve today's issue if we could loose  the requirement (1): all communication between components is through objects should be persistent: 

1) Rate limit on checkpointing NodeStatus (#5953): this causes Daniel's concern on apiserver replication / sharding.
2) Kubelet posts NodeStatus to NodeController as ping message,
NodeController process those messages, and post a comprehensive one to APIserver
3) Similar to 2), but introducing a new ping message. Kubelet sends ping message to NodeController regularly, and only post NodeStatus on change.

**Comment by roberthbailey** [comment]
> 3) Similar to 2), but introducing a new ping message. Kubelet sends ping message to NodeController regularly, and only post NodeStatus on change.

If we don't want the master to initiate contact with the Kubelet and we move to this model, I'd suggest adding a control channel where the master can respond to a ping message with a request for the full NodeStatus. This will allow the master to have the ability to more proactively get current state without needing to change anything in the Kubelet.

**Comment by davidopp** [comment]
I'd prefer that we discuss details of our current NodeStatus problems elsewhere. This proposal is intended to be generic.

I don't think "changing the characteristics of our storage backend" is going to solve the problem that we force every message between Kubernetes components to hit disk. Sure, we can run our key-value store on a huge distributed SSD-based cluster to improve performance but why? Some data really *is* ephemeral and is just a message between two components, not meant for public scrutiny or long-term storage. Why build a crazy-complicated storage system to store things that have no reason to be stored in the first place, instead of just identifying the things that don't need to be stored, and not storing them? And why require a full API sign-off for changing the way two components communicate with each other, if it's not meant for public consumption?

I do think API evolution is related. If you have something that is just an ephemeral message between two components and you can guarantee that the components will be updated atomically, then you don't have to worry about API evolution because you can just restart them together. Even if you can't guarantee atomic update, the fact that they don't use persistent objects for communication means you at least don't have to worry about rewriting on-disk data formats for the objects they use for communicating. And because they're internal components, even in the non-atomic case you only have to support one forward/backward version compatibility--that is, knowing that two components will be upgraded "approximately together" makes things simpler than having to support multiple versions on each end.

I think this proposal kills two birds with one stone: it addresses performance problems we know we're going to have, and by distinguishing internal APIs from external APIs it gives us more flexibility in evolving the former (I admit that the benefit is perhaps even more psychological than it is technical).

**Comment by davidopp** [comment]
BTW see #3247 for an example of where someone was going to use events as an inter-component communication mechanism. This issue is going to come up many times and I think we should try to support it in a reasonable way rather than forcing everything to be a full persisted user-facing API object.

**Comment by davidopp** [comment]
> knowing that two components will be upgraded "approximately together" makes things simpler than 
> having to support multiple versions on each end.

For example, you could say that version N of component A will only talk to version N of component B (i.e. no inter-version compatibility). Even if they don't run on the same machine, as long as you upgrade them approximately together, you can have whichever one comes up first just block until it's talking to a compatible peer. So the total downtime is limited to the time skew between upgrading the components on the two machines. This doesn't work when you have to do rolling upgrade (e.g. you can't have NodeController just block until all Kubelets have been upgraded to a new version) but it does work in the case of master components where there is only a single logical instance of each component.

**Comment by davidopp** [comment]
@thockin @erictune

**Comment by bgrant0607** [comment]
Replying piecemeal: I agree we will want some API resources that are stored in separate etcd instances, or aren't necessarily stored in etcd, or potentially not persisted to disk at all. I've insisted that resource usage collected from nodes be kept out of core API objects for that exact reason:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/resources.md#usage-data

**Comment by bgrant0607** [comment]
I am one of the people who thinks we shouldn't hide "internal" APIs:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/DESIGN.md#overview
(last paragraph of section)

But I do think we should group APIs into multiple API prefixes that can be independently versioned: #3806, #635

**Comment by bgrant0607** [comment]
In Borg, we started with a lot of state not stored persistently and then persisted more and more of it over time, as an inter-component communication medium, as an archival medium, and for stability across master restarts and elections.

In Omega, I enforced a single-writer rule for each object. This meant no true write conflicts, but had other consequences, such as more objects. Joins were never fully realized, which is one reason why Kubernetes started with unified objects containing both spec and status.

**Comment by bgrant0607** [comment]
If we had API plugins #991, we could potentially create API endpoints that more directly dispatched to controllers.

**Comment by bgrant0607** [comment]
I agree with @lavalamp about the sharding. We've discussed sharing storage by namespace and/or kind. My model for sharding by node is to move to federated clusters #5223.

**Comment by bgrant0607** [comment]
If we were to create additional API prefix, I'd like to not put all the types into one file #1744.

**Comment by bgrant0607** [comment]
Internally we effectively persist node pings, but they are made more efficient by batching, patching / delta encoding, compression, and other optimizations.

**Comment by bgrant0607** [comment]
Inter-component APIs must be versioned. We want people to extend the system. Moreover, Kubernetes is open source -- it's a toolkit that people will do with as they wish, including using individual components independently, especially Kubelet, which we've talked about forking into a separate project #444. And many people will be running their own clusters -- I doubt we'll be able to make hard assumptions/requirements about upgrade procedures.

**Comment by davidopp** [comment]
Please don't get hung up on my comment about sharding. My point was just that the approach I described works even in the face of sharding. If you shard etcd rather than API server, it's even easier. I intentionally chose the hardest case (sharding API server) to show that this technique works regardless of the architecture of the master.

**Comment by thockin** [comment]
On Thu, Apr 2, 2015 at 5:14 PM, David Oppenheimer
<notifications@github.com> wrote:

> I do think API evolution is related. If you have something that is just an ephemeral message between two components and you can guarantee that the components will be updated atomically, then you don't have to worry about API evolution because you can just restart them together. Even if you can't guarantee atomic update, the fact that they don't use persistent objects for communication means you at least don't have to worry about rewriting on-disk data formats for the objects they use for communicating. And because they're internal components, even in the non-atomic case you only have to support one forward/backward version compatibility--that is, knowing that two components will be upgraded "approximately together" makes things simpler than having to support multiple versions on each end.

I was going to knee-jerk respond to this, but it got me thinking - do
we have any such case?  Kubernetes is (or was initially) touted as a
reference implementation.  Anyone could implement the API and provide
a kube experience to users.  But does that extend to the idea that our
implementation is a set of components with private channels between
them?  Or do we reasonably expect that someone could take our kubelet
IMPLEMENTATION and use it against a different master?  Or vice-versa?

I've been operating on the assumption that yes, we want that.  Maybe
we don't?  even if not, I still don't see how we can change APIs in
dramatically more nimble ways - nothing is REALLY atomic at the
component-upgrade scope.

But maybe we could define 2 layers of APIs - those that are part of
the user experience and those that are part of the implementation.
For example (straw man): how kubelet gets pod status back to the
master is defined by the implementation.  It could be that the master
polls kubelets.  It could be that kubelet pushes.  Our implementation
is that kubelet uses a GRPC channel carrying ROT13 encoded protbuffers
to the master.  The payload has a 2-release compat windown.  If you
want to use our kubelet but not our master, you have to follow that
implementation and our release train.

**Comment by bgrant0607** [comment]
The problems we had with /operations were:
* they weren't namespaced
* they were global across all kinds
* they were in memory in the apiserver such that they couldn't deliver global semantics in the case that apiserver was replicated

If we did support in-memory objects, they'd need to avoid these problems.

I know you don't want to make this just about NodeStatus, but I agree with @smarterclayton that we should consider a virtual/subresource, perhaps not unlike /binding. The apiserver would then do the object merge, making the communication from Kubelet simpler and more efficient (no roundtrips, smaller messages, timestamps recorded by apiserver).

**Comment by davidopp** [comment]
> nothing is REALLY atomic at the component-upgrade scope.

You could say that version N of component A will only talk to version N of component B (i.e. no inter-version compatibility). Even if they don't run on the same machine, as long as you upgrade them approximately together, you can have them block once they detect they're talking to an incompatible peer (version number, e..g N, could go in every message). So the total downtime is limited to the time delay between upgrading the components on the two machines, and you can avoid forward/backward compatibility of the protocol they use for communicating. (Assuming they communicate using the new kind of objects I described.)

**Comment by smarterclayton** [comment]
A small counter argument to disk being bad - we're getting close to being able to batch commit in etcd at the master level, with both the wal and the raft log being able to coalesce multiple changes into single commits.  Then it really becomes a throughput and contention problem (where contention can be addressed in other ways).  The principle of your argument is sound, but it's a large conceptual change to move suddenly to memory only for a class of problems (complicating clustering).  I'd like to focus on more clearly separating the api for the use case first - the backing mechanism can be a different argument.

**Comment by smarterclayton** [comment]
> On Apr 3, 2015, at 2:46 AM, Tim Hockin <notifications@github.com> wrote:
> 
> On Thu, Apr 2, 2015 at 5:14 PM, David Oppenheimer
> <notifications@github.com> wrote:
> 
> > I do think API evolution is related. If you have something that is just an ephemeral message between two components and you can guarantee that the components will be updated atomically, then you don't have to worry about API evolution because you can just restart them together. Even if you can't guarantee atomic update, the fact that they don't use persistent objects for communication means you at least don't have to worry about rewriting on-disk data formats for the objects they use for communicating. And because they're internal components, even in the non-atomic case you only have to support one forward/backward version compatibility--that is, knowing that two components will be upgraded "approximately together" makes things simpler than having to support multiple versions on each end.
> 
> I was going to knee-jerk respond to this, but it got me thinking - do
> we have any such case? Kubernetes is (or was initially) touted as a
> reference implementation. Anyone could implement the API and provide
> a kube experience to users. But does that extend to the idea that our
> implementation is a set of components with private channels between
> them? Or do we reasonably expect that someone could take our kubelet
> IMPLEMENTATION and use it against a different master? Or vice-versa?

For practical reasons this is what Openshift is doing in some cases.  I'd like a world where you can run our master (kube master + our services) against a set of Kubelet enabled nodes with some version skew. I think tight coupling between the kubelet and master is bad mostly due to this being a distributed system that will see drift over time, and the operational capabilities required to manage this at scale may necessitate several versions of kubelets being alive on the cluster at any one time.

**Comment by bgrant0607** [comment]
Imagine this scenario: Someone starts a Kubernetes cluster and waits 6 months before upgrading anything. It's running their production workloads. They decide they want to upgrade to fix a security problem. They're going to want to perform a rolling update of Kubelets (talking to the old master components) while the cluster is operational, and then upgrade the master. What happens if the new version breaks their workload? In the case that they have to upgrade the whole cluster atomically, they're going to be screwed.

**Comment by smarterclayton** [comment]
+1 to this, especially upgrade.

**Comment by zmerlynn** [comment]
Just to chime in here for a second: Unless we're moving off of what was
discussed at the last meetup and written up in #4855, they're upgrading
master components first, then kubelets. (We've had a few different
conversations now on preventing the master from lighting features prior to
all kubelets being ready as well, so there's some interlock there, too.)

(We discussed the other direction, and they both have tradeoffs. None of
that tooling is done, so it's easy enough to flip around, and that
discussion is probably better done on that issue.)

Imagine this scenario: Someone starts a Kubernetes cluster and waits 6
months before upgrading anything. It's running their production workloads.
They decide they want to upgrade to fix a security problem. They're going
to want to perform a rolling update of Kubelets (talking to the old master
components) while the cluster is operational, and then upgrade the master.
What happens if the new version breaks their workload? In the case that
they have to upgrade the whole cluster atomically, they're going to be
screwed.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/6363#issuecomment-89305187>
.

**Comment by bgrant0607** [comment]
@zmerlynn Sorry, I missed that meetup discussion. I'll take a look.

Anyway, I do agree that there has to be SOME bound on version skew. At some point, we should tell users to turn up a new cluster and migrate their workloads to it, potentially gradually scaling up the new cluster and scaling down the old one. Users will want such workload migration tools for other scenarios, also.

**Comment by bgrant0607** [comment]
Also "in memory" != "not accessible from replicated apiservers". How about etcd on tmpfs?

What's more interesting is whether "in memory" == "not watchable", given that the premise is that the data would be rapidly changing. If watch weren't supported, that would enable more implementation options (e.g., redis?).

**Comment by lavalamp** [comment]
I think it's easiest to reason from multiple distinct concrete cases than make a decision in the abstract.

So let's take NodeStatus. In this case, I agree with @davidopp that there's no reason for kubelet to post the entire node status every time. I can see splitting that into two RPCs, one with the entire status, and one being just a heartbeat, with apiserver filling in the time and doing the write to etcd. We could do this today and I would even call it an improvement.

What we can't do today is store the status ephemerally. This is because we rely on etcd for implementing watch. So imagine we replace etcd, for node status (+ similar things) only, with a component that does in-RAM-only storage/watch for objects. (Let's say we're willing to take a few minutes of inaccuracy while this component warms up after a restart.) This is as far down the road towards @davidopp's proposal as I personally would like to go. (An faster way to do this is etcd on tmpfs, as Brian says immediately above, now that I see his comment.)

Benefits of this:
* It forces us to clean up our storage interface in apiserver.
* We have to change little other apiserver code (after we have subobjects, anyway).
* The choice of whether to persist or not persist status becomes an operational choice.

This doesn't let us do away with our "heavyweight process" around API objects, but I remain unconvinced that that is possible or desirable.

**Comment by brendandburns** [comment]
Ok, there are a bunch of different topics in this thread, so I will try to give my thoughts on a bunch of them.

## Ephemeral Objects
We should absolutely do this.  I know people have heard this before from me, but my over-arching design philosophy is:

*if you can re-calculate it from observing the world, store it in memory, don't store it persistently*

We have moved away from this with NodeStatus, which is really too bad in my opinion, and also has performance problems, as we are seeing with the etcd thrash.

However, we should still have appropriate versioning for these objects, for reasons of non-atomic rollout, random clients, etc.  I think API versioning is different than durability of storage, I'll talk about API versioning next.

## API Versioning
In order to move quickly, while at the same time supporting existing users, we need to be careful about how we manage our APIs.  I propose that we have three api tracks
   * stable - e.g. "1.0" this API is locked and un-changing.  We may make additive changes, but we will never break semantic or syntactic compatability.  We expect most users to use this API.
   * beta - e.g. "v2beta1" this API is the development candidate for version "2.0"  we will work very hard to ensure that there are no breaking changes, but in exceptional cases they may happen.  We expect some users to use this API
  * experimental - e.g. "dev" this is the API under heavy current development, it churns, breaks, etc.  We will disable this API in GKE and in point releases (e.g. 1.1).  We do not expect users to use this API, and if they do, they're in for pain.

Concretely, all initial API development should be done in experimental.  This enables developers to iterate their APIs quickly, without requiring massive investments in compatibility.  It enables them to make mistakes, large changes, and refactors with out affecting end users.  These APIs are best thought of as being under construction.

Eventually, we expect that the APIs will migrate from experimental into beta, at which point the rules are much stronger, as we expect changes to impact users.  Finally, when we are ready, we will cut a new major API version.  These APIs should be thought of as *preview*

## Core Components and Experimental APIs
It's ok if core components use bits of that 'experimental' api while they're under construction, so long as they fail back to something sensible if it's not there (or not what they expect).  This is essential since components and APIs are going to iterate forward together, we need to have speed and agility for both, but we need them to do something sensical when presented with a cluster where experimental is not available.  This can either be done dynamically by introspecting the API, or by flags on the binary.

## Plugin APIs and Schema objects
Eventually we want to support arbitrary user-defined APIs, this will involve making a "Schema" object a first-order k8s object, and allowing users to define schemas that themselves define new object classes.  Objects created this way are inherently outside of the versioning scope of the kubernetes API itself.

**Comment by davidopp** [comment]
That all sounds reasonable to me.

On the subject of APIs, I think having a mechanism for quickly evolving APIs (whether it's the "experimental" API class or plugin APIs) will help not only our own velocity, but will also encourage external developers to contribute stuff upstream rather than having to fork the codebase.

I guess my observation here is that you can only reasonably pick two of these three: (1) design a system out of lots of little components, (2) say all inter-component interactions are part of the API, (3) not have a way to quickly iterate the API. I was suggesting to relax (2), but addressing (3) also solves the problem.

**Comment by smarterclayton** [comment]
We have lots of api infrastructure right now, but not yet enough api automation.  Continuing to improve those pieces will reduce the cost to iterate the api.  We've been better at iterating on the v1beta3 api in an experimental more than we were before.  Some things we should improve:

* partitioning the APIs into groups
* automating/reducing boilerplate around creating our objects
* making it easy to enable APIs via feature flags (like v1beta3)
* more common code for creating and testing api objects 
* better visualization of existing api model objects (swagger is not good enough right now - we need our objects to be visualizable in all their complexity)

**Comment by gmarek** [comment]
I generally agree with David on the API point. I don't like idea of exposing intra cluster API in exactly the same way, and with the same guarantees as "user facing" one.

**Comment by erictune** [comment]
Agree with @smarterclayton that partitioning APIs into groups will help. Those groups should be able to evolve at different speeds.  I can see Node-to-apiserver comms being in one group, and pods/services/rcs in another.  Then, if node-to-apiserver comms changes radically, then it is a headache for the getting-started-guide maintainers, and the cluster admins, maybe.  But it won't invalidate anyone's pod configs.  That seems quite useful.

I don't think David's ephemeral resources idea fits well with the api groups, though.
For example, a node resource may have persistent data in its NodeSpec and ephemeral data in its NodeStatus.  I can't envision the NodeStatus being in one API group and the NodeSpec in another.  That just seems like a lot of groups.   The groups need to be broad, or else customers have to worry about the cross-product of api group versions.

**Comment by bgrant0607** [comment]
API prefixes SGTM. We have lots of use cases for /experimental, such as auto-scaling:
https://github.com/GoogleCloudPlatform/kubernetes/pull/5492#issuecomment-86730490

**Comment by lavalamp** [comment]
API prefixing/grouping SGTM.

I still maintain that storing things ephemerally should be handled by switching out the storage component, NOT by writing different code paths in apiserver. Too many things break otherwise.

**Comment by smarterclayton** [comment]
Agree

----- Original Message -----

**Comment by davidopp** [comment]
Can you explain what you mean "storing things ephemerally should be handled by switching out the storage component"? I think the user experience we want is that objects can be defined as ephemeral or persistent on a per-object basis. It sounds like maybe you're talking about a different axis (i.e. how that would be implemented)?

**Comment by lavalamp** [comment]
>I think the user experience we want is that objects can be defined as ephemeral or persistent on a per-object basis. It sounds like maybe you're talking about a different axis (i.e. how that would be implemented)?

I think the user experience should be that users never notice the difference except in the case that some component restarts. (specifically, watch/list/update/delete/get should all work EXACTLY the same.)

I think the developer experience should be changing some config in e.g. pkg/master that hooks up a different etcd cluster (or different storage component entirely) to that particular resource. The developer experience should emphatically NOT be writing something like our old pod_cache.

**Comment by davidopp** [comment]
Sorry, when I said "user experience" I meant "developer experience."

I think we are agreeing -- I think it should be transparent to the developer whether a particular object is backed by etcd or only by memory. The interface should be the same. Of course the developer needs to be aware of which objects are ephemeral, so that they can be sure to store the state they need for recovery in persistent objects.

**Comment by davidopp** [comment]
Another example of the kind of thing that is difficult/impossible to do because of the requirement that inter-component communication must be via API objects: batching object mutations from clients (scheduler, kubelet, controller manager, etc.) to the API server into a single message. @yujuhong pointed out that today if 30 pods start at the same time on the same kubelet, we do 30 Pod writes. There is presumably some efficiency to be gained by sending a single message to the API server containing all 30 writes. The API server could then turn this into 30 object updates locally.

**Comment by smarterclayton** [comment]
Why is that a problem?

POST /pods {"kind": "PodList"}

Done.

**Comment by derekwaynecarr** [comment]
Assuming it made no transactional commitment, sure, but I think if we start supporting a post of a list, we need a sensible way to give errors if some items in that list fail.

**Comment by derekwaynecarr** [comment]
Or we can do that when etcd has multi object tx support.

**Comment by smarterclayton** [comment]
We have that with causes :). Already thought about it.

**Comment by derekwaynecarr** [comment]
Need to review that I guess.  Where are we posting LIST resources today so I can reference?

**Comment by smarterclayton** [comment]
Nowhere.  The principle is sound though.  We can also expose bulk endpoints if need be when the time comes.

**Comment by bgrant0607** [comment]
I don't object to bulk endpoints, but what efficiency do we hope to gain by it? We definitely need to profile/trace before we do that. I don't believe that it will help, and it could even hurt by serializing operations that could otherwise be performed in parallel. In order to be of benefit, batching would need to reduce the amount of data sent over the wire and/or reduce decoding and other work in the apiserver. 

If etcd supported batching, that could help amortize transaction overheads, but we could send mutations in batches without changing our API.

**Comment by smarterclayton** [comment]
One benefit would be using speedy on larger chunks - probably see much higher compression (keys and common values) for bulk updates than singletons.

**Comment by gmarek** [comment]
What do we want to do with this? It seems that some of things brought up are already implemented, but not all issues mentioned in the first entry are solved.

@bgrant0607 @davidopp @smarterclayton

**Comment by davidopp** [comment]
Let's keep this issue open for future discussion.

**Comment by wojtek-t** [comment]
>
Let's keep this issue open for future discussion.

+1

**Comment by bgrant0607** [comment]
Update:

* The initial experimental API is in, and improvements are in progress (e.g., #13388).
* Full-blown API groups are in progress, which basically permit sparsely populated API group-versions to enable different APIs to evolve independently, additional discovery mechanisms, and more modular code: #10009, #12951
* Third-party CRUD APIs are in progress: #11781, #13000
* Versioned client is underway: #4874, #13098

Unresolved issues remain (in no particular order):

* Non-persistent replicated (HA-friendly) resources, such as for status, Events, Pod/Node Metrics, and resource predictions: storage is now separate from registry. We don't have experience with a non-etcd replication mechanism. We could try an in-memory etcd cluster, or could try moving to Redis or another in-memory caching system.

* Reusable API client and server frameworks, API federation, state sharding, and [proxy offload](https://github.com/kubernetes/kubernetes/issues/3481#issuecomment-74937184): We don't want to link all our resources into a single binary eventually, and don't want a single data path. It doesn't scale and hampers evolution. In addition to making apiserver and the basic client generic and reusable (or switching to a new API platform), we need a federation strategy (e.g., using an [API gateway layer](https://getkong.org)). We need to retain the ability for resources to cross-reference each other and for kubectl and other clients to interact with all configuration and status resources in a generic manner, however. Among other use cases, we'd like to use the API framework in Kubelet #2098 and for resource predictions #12472. 

* Internal APIs: Even with a reusable API framework, people would like it to be easier to develop APIs, by relaxing requirements on API versioning, backward compatibility, documentation, conventions, etc. Out of necessity, we do use a number of different API mechanisms to communicate with "external" dependencies, such as Heapster, InfluxDB, etcd, Docker, and cAdvisor. Except for cAdvisor and Heapster, the others are fairly isolated (completely wrapped by a single system component) and we're developing more comprehensive abstractions to hide them completely over time, such that they are pluggable at the code level (e.g., Cassandra rather than etcd or Rocket instead of Docker). However, the more different mechanisms and conventions we use, the more times we need to re-solve the same problems: authentication and authorization, backward compatibility, client libraries, server infrastructure, defaulting and validation, API discovery, HA, performance, etc. Additionally, in an open-source, extensible toolkit, the notion of "private" APIs is questionable. Someone else is going to use them. And, we also don't have direct control over the cluster upgrade process and frequency. There may be APIs that don't have quite the same requirements, such as use as a declarative configuration schema -- [the "simple REST" convention](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md) should work for those cases. We could explore alternatives, but shouldn't use too many different mechanisms and approaches.

* Extension hooks: We have requests for HTTP-based extensions, such as for the cloudprovider #10503 and scheduler #11470, which are a special case of internal APIs.

**Comment by bgrant0607** [comment]
It's also the case that unstable, non-backward-compatible APIs are hostile to extensibility. That's tantamount to declaring that nobody else in the entire Kubernetes community should be able to extend or replace the client or server without replacing/forking both. That's contradictory to our goals and design principles. To me, that suggests the communication point and/or API abstraction were not chosen correctly.

**Comment by davidopp** [comment]
> That's tantamount to declaring that nobody else in the entire Kubernetes community should be able to extend or replace the client or server without replacing/forking both.

To be clear, what I'm proposing is that we be able to create components that offer only an "internal-only" API. A component that exports an "internal-only" API would be part of the system that people aren't expected to extend or replace independently. (Of course, they could submit PRs that we would upstream into these components.) We can put restrictions on these components like: these API objects are never persisted, and the component must only communicate with a client if the client is of an expected version.

I agree that you absolutely would have to upgrade the component and its client component(s) together; this is what relieves you from having to worry about forward/backward compatibility. The upgrade doesn't have to be atomic, but the component would refuse to talk to a client until the client is upgraded.

You could argue that wanting to be able to quickly evolve an API in incompatible ways is a sign that "the communication point and/or API abstraction were not chosen correctly." But it's often hard to choose these kinds of things correctly up-front; it's useful to have a balance between "prototype and iterate" and making sure you get it right the first time.

I think this is basically a philosophical argument about "in an open-source, extensible toolkit, the notion of "private" APIs is questionable." Clearly there are private interfaces within components, for example the interface between controller manager and controllers, or the way you write a plugin (for our various types of plugins). I'm just suggesting that we also be able to have private interfaces between components, in some limited situations.

**Comment by davidopp** [comment]
@bgrant0607 and I are going to discuss this further offline.

**Comment by davidopp** [comment]
But to clarify one thing I said earlier -- the mechanism I'm describing isn't really for "prototype and iterate." Having alpha API versions and the experimental API prefix already give you that capability. It's about components that you want to have an private/internal API "forever."

**Comment by bgrant0607** [comment]
Every time we create such interdependencies, they also make Kubernetes harder to deploy and upgrade. Many people don't use our /cluster implementation.

We also discussed offline that we could do what we've suggested with the scheduler and its proposed extension API: That when we broke compatibility, WE would fork both sides of the API, so that we wouldn't break users that were dependent on one side or the other.
